presto-hive/src/test/java/com/facebook/presto/hive/TestHiveSplitSource.java

```diff
@@ -319,6 +318,42 @@ public void testPreloadSplitsForGroupedExecution()
         }
     }
 
+    @Test
+    public void testRewind()
+    {
+        HiveSplitSource hiveSplitSource = HiveSplitSource.bucketedRewindable(
+                SESSION,
+                "database",
+                "table",
+                TupleDomain.all(),
+                10,
+                new DataSize(1, MEGABYTE),
+                new TestingHiveSplitLoader(),
+                EXECUTOR,
+                new CounterStat());
+        for (int i = 0; i < 10; i++) {
+            hiveSplitSource.addToQueue(new TestSplit(i, OptionalInt.of(0)));
+            assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), i + 1);
+        }
+        hiveSplitSource.noMoreSplits();
+
+        // Rewind when split is not retrieved.
+        hiveSplitSource.rewind(new HivePartitionHandle(0));
+        assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), 10);
+
+        // Rewind when split is partially retrieved.
+        assertEquals(getSplits(hiveSplitSource, OptionalInt.of(0), 5).size(), 5);
+        assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), 5);
+        hiveSplitSource.rewind(new HivePartitionHandle(0));
+        assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), 10);
+
+        // Rewind when split is fully retrieved
+        assertEquals(getSplits(hiveSplitSource, OptionalInt.of(0), 10).size(), 10);
+        assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), 0);
+        hiveSplitSource.rewind(new HivePartitionHandle(0));
+        assertEquals(hiveSplitSource.getBufferedInternalSplitCount(), 10);
```

Comment:
Thanks for adding additional tests. I would also try to add test for interleaving rewinds

* Rewind bucket 0, test that other buckets are not rewinded
* Rewind bucket 1, test that other buckets are not rewinded
* etc..